### ISSUE TITLE
Zero dilations needs spec clarification

### ISSUE THREAD

**Issue by fdwr** [webmachinelearning/webnn#444]
The [spec](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-conv2d) should clarify 0 dilations behavior as explicitly legal or not. Zero dilations are mathematically sound (it repeats the input value kinda like broadcasting and then dot products it with each filter value), and it is valid in ONNX and DirectML. So rejecting it will make it harder for higher level callers to implement their model. (what about XNNPack though? 🤔)

### Chromium WebNN
Currently rejects it - https://chromium-review.googlesource.com/c/chromium/src/+/4607062 :

```
  if (dilations.height == 0 || dilations.width == 0) {
    return base::unexpected("All dilations should be greater than 0.");
  }
```

### ONNX Runtime library

Allows it. Each input X tensor element is broadcasted to the filter size (1,1,1,5) then multiplied with each W filter element:

    // 2D convolution with 1D convolution kernel, 2 batches and VALID padding.
    {
      "op_type": "Conv",
      "auto_pad": "VALID",
      "strides": [1,1],
      "dilations": [0,0],
      "kernel_shape": [1,5],
      "X": {"dims": [2,1,1,10], "value": [[[[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]]], [[[10, 11, 12, 13, 14, 15, 16, 17, 18, 19]]]]},
      "W": {"dims": [1,1,1,5], "value": [1, 2, 3, 2, 1]},
      "B": [0],
      "Y": [[[[0, 9, 18, 27, 36, 45, 54, 63, 72, 81]]], [[[90, 99, 108, 117, 126, 135, 144, 153, 162, 171]]]]
      "T": "float32"
    }

## PyTorch library

Does not support it.

## TensorFlow library

Does not support it.

## XNNPack backend

Unknown ❔.

## DirectML backend

Supports it.

My vote is to allow it in WebNN as WebNN is a conversion target and lower level, unless there are backends that cannot support it like XNNPack, neither directly nor via decomposition. The fact that PT and TF don't support it just means that they won't need it.

**Comment by wacky6** [comment]
🤔

So `conv2d(kernel_size=(n,m), kernel_weight=ndarray(size=(n,m))` is effectively:
`conv2d(kernel_size=(1,1), kernel_weight=sum(ndarray(size(n,m)))` ?

FYI, JAX doesn't allow `dilation==0` ([code](https://github.com/google/jax/blob/main/jax/_src/lax/lax.py#L4675)). Looks like matlab doesn't support it [as well](https://www.mathworks.com/help/deeplearning/ref/nnet.cnn.layer.convolution2dlayer.html#:~:text=specified%20as%20a%20vector%20%5Bh%20w%5D%20of%20two%20positive%20integers).

Do we know any models that are using this? I guess most models today don't (given popularity of torch/tf/jax among researchers).

**Comment by fdwr** [comment]
@wacky6 : Thanks for the investigation and direct code links 🔎. This artificial limitation across multiple libraries is curious, given zero dilations are mathematically sound - all the equations still work fine with no special cases or logic needed. Maybe nobody thought it was valid because zero is harder to think about and feels like it would collapse or perform a division by zero somewhere (it doesn't). I wager XNNPack might even just work if that validation line was removed. 

> So conv2d(kernel_size=(n,m), kernel_weight=ndarray(size=(n,m)) is effectively:
conv2d(kernel_size=(1,1), kernel_weight=sum(ndarray(size(n,m))) ?`

Exactly - you understand it correctly. ✅

> Do we know any models that are using this?

I do not yet. I was just thinking during code review about the cases where such a model might already exist how we would support that via the WebNN API.
